In [1]:
import sklearn
print(sklearn.__version__)

1.0


In [2]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# call and read digits dataset

digits = load_digits()

print(dir(digits))
print(digits.target_names)
print(digits.DESCR)

# assign data for later use

digits_feature = digits.data
digits_label = digits.target

['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']
[0 1 2 3 4 5 6 7 8 9]
.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bit

In [3]:
from sklearn.model_selection import train_test_split

# split data for model training and model testing

data_train, data_test, label_train, label_test = train_test_split(digits_feature, digits_label, test_size = 0.2, random_state=7)
                                                     
print('data_train 개수: ', len(data_train),', data_test 개수: ', len(data_test))
print(data_train.shape, label_train.shape)
print(data_test.shape, label_test.shape)

data_train 개수:  1437 , data_test 개수:  360
(1437, 64) (1437,)
(360, 64) (360,)


In [4]:
# import models

from sklearn.tree import DecisionTreeClassifier 
from sklearn.ensemble import RandomForestClassifier  
from sklearn import svm  
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression 

# import metrics

from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

---
## Trying Decision Tree model


In [5]:
decision_tree = DecisionTreeClassifier(random_state=32) 
decision_tree.fit(data_train, label_train)
label_prediction = decision_tree.predict(data_test)

print(confusion_matrix(label_test, label_prediction))
print(classification_report(label_test, label_prediction))

[[42  0  0  1  0  0  0  0  0  0]
 [ 0 34  3  1  0  1  1  0  0  2]
 [ 0  0 33  2  0  0  1  1  2  1]
 [ 0  1  0 31  0  0  0  0  1  1]
 [ 0  0  1  0 35  0  0  0  1  0]
 [ 0  1  0  0  0 27  0  0  0  0]
 [ 0  0  0  0  2  0 26  0  0  0]
 [ 0  0  0  1  2  1  0 27  0  2]
 [ 0  5  4  1  1  0  3  0 28  1]
 [ 0  1  1  2  2  1  0  0  0 25]]
              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       

---
## Trying Randome Foreest model

In [6]:
# trying Random Forest model

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(data_train, label_train)
label_prediction = random_forest.predict(data_test)

print(confusion_matrix(label_test, label_prediction))
print(classification_report(label_test, label_prediction))

[[42  0  0  0  1  0  0  0  0  0]
 [ 0 42  0  0  0  0  0  0  0  0]
 [ 0  0 40  0  0  0  0  0  0  0]
 [ 0  0  0 34  0  0  0  0  0  0]
 [ 0  0  0  0 37  0  0  0  0  0]
 [ 0  0  0  0  0 27  0  0  0  1]
 [ 0  0  0  0  1  0 27  0  0  0]
 [ 0  0  0  0  0  0  0 32  0  1]
 [ 0  3  0  0  1  1  0  2 36  0]
 [ 0  0  0  0  0  2  0  0  0 30]]
              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       0.93      1.00      0.96        37
           5       0.90      0.96      0.93        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       1.00      0.84      0.91        43
           9       0.94      0.94      0.94        32

    accuracy                           0.96       360
   macro avg       

---
## Trying Support Vector Machine

In [8]:
# trying Support Vector Machine

svm_model = svm.SVC()
svm_model.fit(data_train, label_train)
label_prediction = svm_model.predict(data_test)

print(confusion_matrix(label_test, label_prediction))
print(classification_report(label_test, label_prediction))

[[43  0  0  0  0  0  0  0  0  0]
 [ 0 42  0  0  0  0  0  0  0  0]
 [ 0  0 40  0  0  0  0  0  0  0]
 [ 0  0  0 34  0  0  0  0  0  0]
 [ 0  0  0  0 37  0  0  0  0  0]
 [ 0  0  0  0  0 28  0  0  0  0]
 [ 0  0  0  0  0  0 28  0  0  0]
 [ 0  0  0  0  0  0  0 33  0  0]
 [ 0  2  0  0  0  1  0  0 40  0]
 [ 0  0  0  0  0  1  0  0  0 31]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       

---
## Trying Stochastic Gradient Descent model

In [9]:
sgd_model = SGDClassifier()
sgd_model.fit(data_train, label_train)
label_prediction = sgd_model.predict(data_test)

print(confusion_matrix(label_test, label_prediction))
print(classification_report(label_test, label_prediction))

[[41  0  0  0  1  0  0  1  0  0]
 [ 0 35  0  2  0  0  0  0  4  1]
 [ 0  0 39  1  0  0  0  0  0  0]
 [ 0  0  0 32  0  0  0  1  1  0]
 [ 0  0  0  0 37  0  0  0  0  0]
 [ 0  0  0  0  0 28  0  0  0  0]
 [ 0  1  0  0  0  0 26  0  1  0]
 [ 0  0  0  1  0  0  0 32  0  0]
 [ 0  4  0  1  1  0  0  0 37  0]
 [ 0  1  0  3  0  2  0  0  1 25]]
              precision    recall  f1-score   support

           0       1.00      0.95      0.98        43
           1       0.85      0.83      0.84        42
           2       1.00      0.97      0.99        40
           3       0.80      0.94      0.86        34
           4       0.95      1.00      0.97        37
           5       0.93      1.00      0.97        28
           6       1.00      0.93      0.96        28
           7       0.94      0.97      0.96        33
           8       0.84      0.86      0.85        43
           9       0.96      0.78      0.86        32

    accuracy                           0.92       360
   macro avg       

---
## Trying Logistic Regression

In [22]:
logistic_model = LogisticRegression(max_iter=2000) # 최대 반복횟수 증가
logistic_model.fit(data_train, label_train)
label_prediction = logistic_model.predict(data_test)

print(confusion_matrix(label_test, label_prediction))
print(classification_report(label_test, label_prediction))

[[43  0  0  0  0  0  0  0  0  0]
 [ 0 40  0  0  0  0  0  0  1  1]
 [ 0  0 40  0  0  0  0  0  0  0]
 [ 0  0  0 33  0  0  0  1  0  0]
 [ 0  0  0  0 37  0  0  0  0  0]
 [ 0  0  0  0  0 27  0  0  1  0]
 [ 0  0  0  0  0  0 27  0  1  0]
 [ 0  0  0  1  0  0  0 32  0  0]
 [ 0  2  1  0  0  4  0  1 35  0]
 [ 0  0  0  1  0  3  0  0  0 28]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       1.00      1.00      1.00        37
           5       0.79      0.96      0.87        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       0.92      0.81      0.86        43
           9       0.97      0.88      0.92        32

    accuracy                           0.95       360
   macro avg       

## 모델을 평가해 보기
숫자를 인식하는 문제의 경우, False Positive와 False Negative 중 특별히 더 중요한 것이 없는데, 이는 숫자의 인식 자체를 실패하는 경우를 제외하면, 특정 숫자의 False Negative가 다른 숫자의 False Postiive와 동일한 결과를 낳기 때문이다. 따라서 정밀도와 재현율 중 어느 것이 더 중요한 표지라고 말할 수 없으므로, 둘의 조화평균인 F1 Score를 평가기준으로 삼는 것이 타당하다. 한편, 본 시행에서 각 항목별 support가 크게는 1.5배까지 차이나므로, F1 score의 macro avg가 아닌 weighted avg 값을 이용해 평가한다.

해당 평가기준에 따라 5가지 모델을 비교하면 다음과 같다.

|모델명|w_avg of F1 Score|순위|
|---|---|---|
|의사결정나무|0,85|5|
|랜덤포레스트|0.96|2|
|SVM|0.99|1|
|SGD|0.92|4|
|로지스틱|0.95|3|

따라서 해당 모델 중 SVM 모델이 가장 우수하며, 의사결정나무는 